snowfall/decoding/nbest_rescore.py

```diff
@@ -0,0 +1,112 @@
+# Copyright (c)  2021  Xiaomi Corporation (authors: Wei Kang)
+
+import k2
+import torch
+
+
+def get_best_matching_stats(keys: k2.Nbest, queries: k2.Nbest,
+                            max_order: int) -> torch.Tensor:
+    '''Get best matching stats on query positions.
+
+    Args:
+      keys:
+        The nbest after doing second pass rescoring.
+      queries:
+        Another nbest before doing second pass rescoring.
+      max_order:
+        The maximum n-gram order to ever return by `k2.get_best_matching_stats`
+
+    Returns:
+      A tensor with the shape of [queries.fsa.num_elements, 5], each row
+      contains the stats (init_score, mean, var, counts_out, ngram_order)
+      of the token in the correspodding position in queries.
+    '''
+    assert keys.shape.dim0() == queries.shape.dim0(), \
+        f'Utterances number in keys and queries should be equal : \
+         {keys.shape.dim0()} vs {queries.shape.dim0()}'
+
+    # keys_tokens_shape [utt][path][token]
+    keys_tokens_shape = k2.ragged.compose_ragged_shapes(keys.shape,
+        k2.ragged.remove_axis(keys.fsa.arcs.shape(), 1))
+    # queries_tokens_shape [utt][path][token]
+    queries_tokens_shape = k2.ragged.compose_ragged_shapes(queries.shape,
+        k2.ragged.remove_axis(queries.fsa.arcs.shape(), 1))
+
+    keys_tokens = k2.RaggedInt(keys_tokens_shape, keys.fsa.labels.clone())
+    queries_tokens = k2.RaggedInt(queries_tokens_shape,
+                                  queries.fsa.labels.clone())
+    # tokens shape [utt][path][token]
+    tokens = k2.ragged.cat([keys_tokens, queries_tokens], axis=1)
+
+    keys_token_num = keys.fsa.labels.size()[0]
+    queries_tokens_num = queries.fsa.labels.size()[0]
+    # counts on key positions are ones
+    keys_counts = k2.RaggedInt(keys_tokens_shape,
+                               torch.ones(keys_token_num,
+                                          dtype=torch.int32))
+    # counts on query positions are zeros
+    queries_counts = k2.RaggedInt(queries_tokens_shape,
+                                  torch.zeros(queries_tokens_num,
+                                              dtype=torch.int32))
+    counts = k2.ragged.cat([keys_counts, queries_counts], axis=1).values()
+
+    # scores on key positions are the scores inherted from nbest path
+    keys_scores = k2.RaggedFloat(keys_tokens_shape, keys.fsa.scores.clone())
+    # scores on query positions MUST be zeros
+    queries_scores = k2.RaggedFloat(queries_tokens_shape,
+                                    torch.zeros(queries_tokens_num,
+                                                dtype=torch.float32))
+    scores = k2.ragged.cat([keys_scores, queries_scores], axis=1).values()
+
+    # we didn't remove -1 labels before
+    min_token = -1
+    eos = -1
```

Comment:
I did't remove -1 labels in nbest paths,  if I removed these labels I had to remove the scores in the corresponding positions, I did not find functions could do this thing. 

I thought the remained -1 labels should not be a problem.

The `eos` could be other value, or as a parameter just like the `max_order`.